hera_cal/lst_stack/calibration.py

```diff
@@ -0,0 +1,371 @@
+                # Construct the raw gain and remove nans and infs
+                raw_gain = calibration_parameters[f"A_J{pol}"] ** 2 * (
+                    phase_gains.get(
+                        (ant1, "J" + pol),
+                        np.ones_like(calibration_parameters[f"A_J{pol}"]),
+                    )
+                    * phase_gains.get(
+                        (ant2, "J" + pol),
+                        np.ones_like(calibration_parameters[f"A_J{pol}"]),
+                    ).conj()
+                )
+                raw_gain = np.where(
+                    stack.flags[:, 0, band, polidx], 1.0 + 0.0j, raw_gain[:, band]
+                )
+
+                # Compute smooth gain for each parameter and remove zeros/nans/infs
+                if smooth_gains:
+                    basis = smoothing_functions[bandidx]
+                    bl_gain_here = np.array(
+                        [
+                            np.dot(basis, _fmat.dot(_raw_gain))
+                            for _fmat, _raw_gain in zip(fmats[pol][bandidx], raw_gain)
+                        ]
+                    )
+                else:
+                    bl_gain_here = raw_gain
+
+                bl_gain_here = np.where(
+                    np.isfinite(bl_gain_here), bl_gain_here, 1.0 + 0.0j
+                )
+                bl_gain_here = np.where(
+                    np.isclose(bl_gain_here, 0), 1.0 + 0.0j, bl_gain_here
+                )
+                bl_gain[:, band] = bl_gain_here
+
+            # Calibrate out smoothed gains
+            if calibrate_inplace:
+                stack.data[:, apidx, :, polidx] /= bl_gain
+
+            # Store gains to return
+            if return_gains:
+                gains[(ant1, ant2, pol)] = bl_gain
+
+    if auto_stack:
+        for polidx, pol in enumerate(pols):
+            for apidx, (ant1, ant2) in enumerate(auto_stack.antpairs):
+                bl_gain = np.ones(
+                    (stack.nights.size, stack.freq_array.size), dtype=complex
+                )
+                for bandidx, band in enumerate(inpaint_bands):
+                    # Construct the raw gain and remove nans and infs
+                    raw_gain = calibration_parameters[f"A_J{pol}"] ** 2
+                    raw_gain = np.where(
+                        stack.flags[:, 0, band, polidx], 1.0 + 0.0j, raw_gain[:, band]
+                    )
+
+                    # Compute smooth gain for each parameter and remove zeros/nans/infs
+
+                    if smooth_gains:
+                        basis = smoothing_functions[bandidx]
+                        bl_gain_here = np.array(
+                            [
+                                np.dot(basis, _fmat.dot(_raw_gain))
+                                for _fmat, _raw_gain in zip(
+                                    fmats[pol][bandidx], raw_gain
+                                )
+                            ]
+                        )
```

Comment:
First off, I think we should rephase before we smooth (see smooth_cal for how to do that). Second, I'm wondering if we should smooth antenna gains rather than gain products.

Reply:
I added in rephasing and smoothing antenna gains rather than the antenna gain product